src/analytics/TraceEvent.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { Children, cloneElement, isValidElement, memo, PropsWithChildren, SyntheticEvent } from 'react'
+
+import { sendAnalyticsEvent } from '.'
+import { Event, EventName } from './constants'
+import { ITraceContext, Trace, TraceContext } from './Trace'
+
+type TraceEventProps = {
+  events: Event[]
+  name: EventName
+  properties?: Record<string, unknown>
+  shouldLogImpression?: boolean
+} & ITraceContext
+
+/**
+ * Analytics instrumentation component that wraps event callbacks with logging logic.
+ *
+ * @example
+ *  <TraceEvent events={[Event.onClick]} element={ElementName.SWAP_BUTTON}>
+ *    <Button onClick={() => console.log('clicked')}>Click me</Button>
+ *  </TraceEvent>
+ */
+export const TraceEvent = memo((props: PropsWithChildren<TraceEventProps>) => {
+  const { shouldLogImpression, name, properties, events, children, ...traceProps } = props
+
+  return (
+    <Trace {...traceProps}>
+      <TraceContext.Consumer>
+        {(traceContext) =>
+          Children.map(children, (child) => {
+            if (!isValidElement(child)) {
+              return child
+            }
+
+            // For each child, augment event handlers defined in `events` with event tracing.
+            return cloneElement(
+              child,
+              getEventHandlers(child, traceContext, events, name, properties, shouldLogImpression)
+            )
+          })
+        }
+      </TraceContext.Consumer>
+    </Trace>
+  )
+})
+
+TraceEvent.displayName = 'TraceEvent'
+
+/**
+ * Given a set of child element and event props, returns a spreadable
+ * object of the event handlers augmented with analytics logging.
+ */
+function getEventHandlers(
+  child: React.ReactElement,
+  traceContext: ITraceContext,
+  events: Event[],
+  name: EventName,
+  properties?: Record<string, unknown>,
+  shouldLogImpression = true
+) {
+  const eventHandlers: Partial<Record<Event, (e: SyntheticEvent<Element, Event>) => void>> = {}
+
+  for (const event of events) {
```

Comment:
style: any objection to using a forEach here?

Reply:
For styling preference comments, instead of discussing, we can enforce a formatting rule. No objection to changing this, if there is a formatting rule that supports it.

Reply:
After some research this one seems more context dependent / not lintable so will leave as is